sql/src/main/java/io/crate/execution/engine/aggregation/impl/CollectSetAggregation.java

```diff
@@ -109,4 +109,26 @@ public DataType partialType() {
     public Set<Object> terminatePartial(RamAccountingContext ramAccountingContext, Set<Object> state) {
         return state;
     }
+
+    @Override
+    public boolean isRemovableCumulative() {
+        return true;
+    }
+
+    @Override
+    public Set<Object> removeFromAggregatedState(RamAccountingContext ramAccountingContext,
+                                                 Set<Object> previousAggState,
+                                                 Input[] stateToRemove) {
+        Object value = stateToRemove[0].value();
+        if (value == null) {
+            return previousAggState;
+        }
+        if (previousAggState.remove(value)) {
```

Comment:
Isn't it possible that this value had been occurring multiple times and it would incorrectly be removed here due to the de-duplication of the set?

Reply:
hm good point, I think in the peers case (ordered windows) it wouldn't be a problem, but let me think about the non-ordered windows

Reply:
ok, so this won't work with un-ordered partitions, where peers are not part of the same frame. In order to achieve this we'd need more metadata to pass to the aggregation (like, a count of peers in the frame and decide if we need to remove) or for the aggregation to expose some metadata  (to indicate we should skip removing of elements in the computation of the function if we detect peers in the frame) 

As collect_set is an internal function, not sure if it's worth doing anything about it at this stage. What do you think @mfussenegger  ?

Reply:
I guess it is not worth doing that for now.